## Profitable Apps Project





**Project Description**

Aim of Project is to find mobile apps which are profitable  for app store and google play markets. As Data analyst, I want to enable developers via data driven decisions with respect to kind of apps they build. 

In [1]:
#Opening Apple and Google Datasets using Absolute Paths
from csv import reader
opened_applefile = open('E:\Educational\Python\Datasets\AppleStore.csv', encoding='utf8')
read_apple = reader(opened_applefile)
data_apple = list(read_apple)
apple_rows = data_apple[1:] #Slicing out the header from the dataset
apple_header = data_apple[0]#only Headers of the dataset

opened_googlefile = open('E:\Educational\Python\Datasets\googleplaystore.csv', encoding='utf8')
read_google = reader(opened_googlefile)
data_google = list(read_google)
google_rows = data_google[1:]
google_header = data_google[0]

#calling datasets for check
print(apple_rows[0:2],'\n')
print(google_rows[0:2])


[['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']] 

[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']]


## Creating a function to call data sets

In [2]:
def explore_data(dataset,start,end,rows_cols = True):
    for r in dataset[start:end]:
        print(r,'\n')
    if rows_cols:
        print('Number of rows :', len(dataset))
        print('Number of columns:',len(dataset[0]))

In [3]:
print(google_header,'\n')
explore_data(google_rows,0,2) #rows_cols condition True by default no need to place an argument for it in function

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up'] 

Number of rows : 10841
Number of columns: 13


## Removing Duplicates & Non English/Free Apps
**Data Cleaning**

Reading Kaggle discussion forumn, it was shown that data set is missing a column value. How to determine which row has that missing value in the dataset?

In [4]:
for rows in google_rows:
    Col_len = len(google_header)
    row_len = len(rows)
    if Col_len != row_len:
        print(rows,'\n')
        print('Row with missing column is', google_rows.index(rows))



['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up'] 

Row with missing column is 10472


## Deleting wrong Data

We will compare the row 10472 with headers to understand origin of the data error. Then delete the row. Two steps ofcourse are mutually exclusive.

In [5]:
print(google_header,'\n')
print(google_rows[1],'\n')
print(google_rows[10472])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up'] 

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


As we can see the row 10472 is missing the Category Column, this leads to eroneous alignment for other data points for example Ratings for app in this row is 19. However, Google playstore ratings are only upto 5.

In [6]:
print(len(google_rows))
del google_rows[10472]
print(len(google_rows))

10841
10840


## Removing Duplicates

**PART 1**

**Creating two Seperate lists for Duplicates & Unique values**

Logic: if an element in dataset exists in unique apps already then it is a duplicate. Hence we should add it to duplicate list via append method. 

In [7]:
duplicate_apps = []
unique_apps = []

for apps in google_rows:
    name = apps[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
duplicate_apps[0:10] #some examples of duplicate apps

['Quick PDF Scanner + OCR FREE',
 'Box',
 'Google My Business',
 'ZOOM Cloud Meetings',
 'join.me - Simple Meetings',
 'Box',
 'Zenefits',
 'Google Ads',
 'Google My Business',
 'Slack']

## PART 2
**Removing Duplicates on a Criterion**

As you can see each row with respective duplicate name is copied across the database. Quickly eye balling the Slack app duplicates in dataset we can observe that higher the number of reviews more recent the data should be. Hence rather than removing duplicates randomly, we will only keep entries of duplicates with highest number of reviews whilst remove others.

In [8]:
reviews_max = {}

for e in google_rows:
    name = e[0]
    n_reviews = float(e[3])
    
    if name in reviews_max and reviews_max[name]< n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
#Important to use ELIF here and not ELSE - using else will also take exception of n_reviews.
#We only want to pivot on names

In [9]:
print('Expected Length',len(reviews_max),'\n')
print(len(duplicate_apps),'\n')
print('Actual Length',len(google_rows)-len(duplicate_apps))

Expected Length 9659 

1181 

Actual Length 9659



Now, let's use the reviews_max dictionary to remove the duplicates. For the duplicate cases, we'll only keep the entries with the highest number of reviews. In the code cell below:

* We start by initializing two empty lists, android_clean and already_added.
* We loop through the android data set, and for every iteration:
    * We isolate the name of the app and the number of reviews.
    * We add the current row (app) to the android_clean list, and the app name     (name) to the already_added list if:

        - The number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary; and

        - The name of the app is not already in the already_added list. We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [10]:
google_clean = []
already_added = []

for rows in google_rows:
    name = rows[0]
    n_reviews = float(rows[3])
    
    if reviews_max[name]== n_reviews and name not in already_added:
        google_clean.append(rows)
        already_added.append(name)



Exploring the clean data set, it should give us same dataset length and width as created above in the dictionary. 

In [11]:
explore_data(google_clean,0,4)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'] 

['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up'] 

Number of rows : 9659
Number of columns: 13


**As expected - we have 9659 rows.**

# Removing Non English Apps

**Part One**

Exploring the data set we can see there are many apps which are not in english language hence we will end targeting the wrong audience. Below we will sort the data and segregate english apps from non english.

**Use the Built-in Ord() function to leverage ASCII character standard**

* All english alphabets have ASCII numerical value between 0 and 127

In [12]:
def is_english(string):
    for element in string:
        if ord(element) > 127:
            return False
        
    return True
            

        
    

In [13]:
print(is_english('Instagram '))
print(is_english('a视剧热播'))

print(is_english('Instachat 😜'))
print(is_english('Docs To Go™ Free Office Suite'))
print(ord("剧"))

True
False
False
False
21095


However as you can see, the function is not perfect. Some apps use emojis or symbols in their names. Due to which the existing function check will drop them in selection. One way to limit data loss is by setting a criteria. For example. The apps which have atleast 3 characters outside ASCII range, will be deemed **Non English** . 

# Part Two

In [14]:
def is_english(string):
    non_ascii = 0
    
    for e in string:
        if ord(e)>127:
            non_ascii += 1
    
    if non_ascii >3:
        return False
    else:
        return True
    

In [15]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
True


Although Function is still not but this limits the data loss quite a bit.

# Seperating English & Non English into Seperate Lists Via Defined Function

In [16]:
google_english = []
apple_english = []

for rows in google_clean:
    name = rows[0]
    
    if is_english(name):
        google_english.append(rows)
        

for rows in apple_rows:
    name = rows[1]
    
    if is_english(name):
        apple_english.append(rows)
        
explore_data(google_english,0,2)
print('\n')
explore_data(apple_english,0,2)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

Number of rows : 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'] 

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'] 

Number of rows : 6183
Number of columns: 16


# Isolating the Free Apps



In [17]:
google_free = []
apple_free = []

for rows in google_english:
    price = rows[7]
    
    if price == '0':
        google_free.append(rows)
        

for rows in apple_english:
    price = rows[4]
    
    if price == '0.0':
        apple_free.append(rows)

print(len(google_free), "android free apps")
print(len(apple_free), "ios free apps")



8864 android free apps
3222 ios free apps


# Finding Common Genres in IOS & Andorid App Stores

Pursuing a strategy we need to find an app which is most likely to be used by most number of people on both IOS & Google play stores. 

To do so , we will create few functions which will be applicable to both ios and andorid datasets.

* Function to display Genres by Absolute frequency

* Function to display Genres by percentage proportion of all

In [18]:
#Creating function to genreate absolute frequency table with ascending/desending order flexibility
def abs_frequency(dataset,index,descend = True):
    table = {}
    total = 0
    #altnernate to total counter : sum(list(table.values())) -> this will sum the total occurences in dictionary and be same thing as total counter
    for rows in dataset:
        genre = rows[index]
        total += 1
        if genre in table:
            table[genre] += 1
        else:
            table[genre] = 1
    if descend:
        return sorted(table.items(),key=lambda x:x[1],reverse = True) #to ensure table rolls out in descending order
    else:
        return sorted(table.items(),key=lambda x:x[1],reverse = False) # Ascending order

            


#Creating function to display genres in percentage, ascending/descending flexibility

def display_table(dataset, index, descend = True):
    table = {}
    total = 0
    
    for rows in dataset:
        genre = rows[index]
        total += 1
        if genre in table:
            table[genre] += 1
        else:
            table[genre] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key]/total)*100
        table_percentages[key] = percentage
        
    if descend:
        return sorted(table_percentages.items(),key=lambda x:x[1],reverse=True)
    else:
        return sorted(table_percentages.items(),key=lambda x:x[1],reverse=False)
    
#Important to define x:x[1] as it will always then apply sort on whatever value on index position at 1
#after sort is applied dictionary is changed to type list

# Part three

We start examining the frequency table for genres in both data bases.

In [19]:
display_table(apple_free,11)

[('Games', 58.16263190564867),
 ('Entertainment', 7.883302296710118),
 ('Photo & Video', 4.9658597144630665),
 ('Education', 3.662321539416512),
 ('Social Networking', 3.2898820608317814),
 ('Shopping', 2.60707635009311),
 ('Utilities', 2.5139664804469275),
 ('Sports', 2.1415270018621975),
 ('Music', 2.0484171322160147),
 ('Health & Fitness', 2.0173805090006205),
 ('Productivity', 1.7380509000620732),
 ('Lifestyle', 1.5828677839851024),
 ('News', 1.3345747982619491),
 ('Travel', 1.2414649286157666),
 ('Finance', 1.1173184357541899),
 ('Weather', 0.8690254500310366),
 ('Food & Drink', 0.8069522036002483),
 ('Reference', 0.5586592178770949),
 ('Business', 0.5276225946617008),
 ('Book', 0.4345127250155183),
 ('Navigation', 0.186219739292365),
 ('Medical', 0.186219739292365),
 ('Catalogs', 0.12414649286157665)]

In [20]:
display_table(google_free,1)

[('FAMILY', 18.907942238267147),
 ('GAME', 9.724729241877256),
 ('TOOLS', 8.461191335740072),
 ('BUSINESS', 4.591606498194946),
 ('LIFESTYLE', 3.9034296028880866),
 ('PRODUCTIVITY', 3.892148014440433),
 ('FINANCE', 3.7003610108303246),
 ('MEDICAL', 3.531137184115524),
 ('SPORTS', 3.395758122743682),
 ('PERSONALIZATION', 3.3167870036101084),
 ('COMMUNICATION', 3.2378158844765346),
 ('HEALTH_AND_FITNESS', 3.0798736462093865),
 ('PHOTOGRAPHY', 2.944494584837545),
 ('NEWS_AND_MAGAZINES', 2.7978339350180503),
 ('SOCIAL', 2.6624548736462095),
 ('TRAVEL_AND_LOCAL', 2.33528880866426),
 ('SHOPPING', 2.2450361010830324),
 ('BOOKS_AND_REFERENCE', 2.1435018050541514),
 ('DATING', 1.861462093862816),
 ('VIDEO_PLAYERS', 1.7937725631768955),
 ('MAPS_AND_NAVIGATION', 1.3989169675090252),
 ('FOOD_AND_DRINK', 1.2409747292418771),
 ('EDUCATION', 1.1620036101083033),
 ('ENTERTAINMENT', 0.9589350180505415),
 ('LIBRARIES_AND_DEMO', 0.9363718411552346),
 ('AUTO_AND_VEHICLES', 0.9250902527075812),
 ('HOUSE_AN

We can see that among Free English apps in IOS store almost 58 % are games and then entertainment apps are close to 8 %. Then followed by photography and social media apps. However 'fun' apps in IOS quite large does not imply they will have the most users as well. Demand might not be the same. 

In [21]:
display_table(google_free,1)

[('FAMILY', 18.907942238267147),
 ('GAME', 9.724729241877256),
 ('TOOLS', 8.461191335740072),
 ('BUSINESS', 4.591606498194946),
 ('LIFESTYLE', 3.9034296028880866),
 ('PRODUCTIVITY', 3.892148014440433),
 ('FINANCE', 3.7003610108303246),
 ('MEDICAL', 3.531137184115524),
 ('SPORTS', 3.395758122743682),
 ('PERSONALIZATION', 3.3167870036101084),
 ('COMMUNICATION', 3.2378158844765346),
 ('HEALTH_AND_FITNESS', 3.0798736462093865),
 ('PHOTOGRAPHY', 2.944494584837545),
 ('NEWS_AND_MAGAZINES', 2.7978339350180503),
 ('SOCIAL', 2.6624548736462095),
 ('TRAVEL_AND_LOCAL', 2.33528880866426),
 ('SHOPPING', 2.2450361010830324),
 ('BOOKS_AND_REFERENCE', 2.1435018050541514),
 ('DATING', 1.861462093862816),
 ('VIDEO_PLAYERS', 1.7937725631768955),
 ('MAPS_AND_NAVIGATION', 1.3989169675090252),
 ('FOOD_AND_DRINK', 1.2409747292418771),
 ('EDUCATION', 1.1620036101083033),
 ('ENTERTAINMENT', 0.9589350180505415),
 ('LIBRARIES_AND_DEMO', 0.9363718411552346),
 ('AUTO_AND_VEHICLES', 0.9250902527075812),
 ('HOUSE_AN

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.

# Most Popular Apps by Genre on App Store

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

In [22]:
genres_ios= abs_frequency(apple_free,-5) # this will return a LIST not a dictionary, hence genre has to be defined within the loop

for rows in genres_ios:
    genre = rows[0]
    total = 0
    len_genre = 0
    for app in apple_free:
        genre_app = app[-5]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_rating = total/len_genre
    print(genre,':',avg_rating)

Games : 22788.6696905016
Entertainment : 14029.830708661417
Photo & Video : 28441.54375
Education : 7003.983050847458
Social Networking : 71548.34905660378
Shopping : 26919.690476190477
Utilities : 18684.456790123455
Sports : 23008.898550724636
Music : 57326.530303030304
Health & Fitness : 23298.015384615384
Productivity : 21028.410714285714
Lifestyle : 16485.764705882353
News : 21248.023255813954
Travel : 28243.8
Finance : 31467.944444444445
Weather : 52279.892857142855
Food & Drink : 33333.92307692308
Reference : 74942.11111111111
Business : 7491.117647058823
Book : 39758.5
Navigation : 86090.33333333333
Medical : 612.0
Catalogs : 4004.0


On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together:

In [23]:
for app in apple_free:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5



The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc. Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number.

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

# Most Popular apps by Genre on Play store

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [24]:
display_table(google_free,5)

[('1,000,000+', 15.726534296028879),
 ('100,000+', 11.552346570397113),
 ('10,000,000+', 10.548285198555957),
 ('10,000+', 10.198555956678701),
 ('1,000+', 8.393501805054152),
 ('100+', 6.915613718411552),
 ('5,000,000+', 6.825361010830325),
 ('500,000+', 5.561823104693141),
 ('50,000+', 4.7721119133574),
 ('5,000+', 4.512635379061372),
 ('10+', 3.5424187725631766),
 ('500+', 3.2490974729241873),
 ('50,000,000+', 2.3014440433213),
 ('100,000,000+', 2.1322202166064983),
 ('50+', 1.917870036101083),
 ('5+', 0.78971119133574),
 ('1+', 0.5076714801444043),
 ('500,000,000+', 0.2707581227436823),
 ('1,000,000,000+', 0.22563176895306858),
 ('0+', 0.04512635379061372),
 ('0', 0.01128158844765343)]

One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [25]:
categories_android = abs_frequency(google_free,1)
install_dist = []

for e in categories_android:
    category = e[0]
    total = 0
    len_categ = 0
    for rows in google_free:
        category_app = rows[1]
        if category_app == category:
            n_installs = rows[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            n_installs = float(n_installs)
            total += n_installs
            len_categ += 1
    
    avg_installs = total/len_categ
    print(category,':',avg_installs)
            
#The nested loops follow block logic, you can re build a small case using example on page 12


    

FAMILY : 3695641.8198090694
GAME : 15588015.603248259
TOOLS : 10801391.298666667
BUSINESS : 1712290.1474201474
LIFESTYLE : 1437816.2687861272
PRODUCTIVITY : 16787331.344927534
FINANCE : 1387692.475609756
MEDICAL : 120550.61980830671
SPORTS : 3638640.1428571427
PERSONALIZATION : 5201482.6122448975
COMMUNICATION : 38456119.167247385
HEALTH_AND_FITNESS : 4188821.9853479853
PHOTOGRAPHY : 17840110.40229885
NEWS_AND_MAGAZINES : 9549178.467741935
SOCIAL : 23253652.127118643
TRAVEL_AND_LOCAL : 13984077.710144928
SHOPPING : 7036877.311557789
BOOKS_AND_REFERENCE : 8767811.894736841
DATING : 854028.8303030303
VIDEO_PLAYERS : 24727872.452830188
MAPS_AND_NAVIGATION : 4056941.7741935486
FOOD_AND_DRINK : 1924897.7363636363
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
LIBRARIES_AND_DEMO : 638503.734939759
AUTO_AND_VEHICLES : 647317.8170731707
HOUSE_AND_HOME : 1331540.5616438356
WEATHER : 5074486.197183099
EVENTS : 253542.22222222222
PARENTING : 542603.6206896552
ART_AND_DESIGN : 198

On average - communication genre has most installs as shownn by the distribution generated above. However upon analyzing the category closely we can observer how heavily skewed this category is. As it will be true for most other categories too.

In [26]:
for apps in google_free:
    if apps [1] == 'COMMUNICATION' and (apps[5] == '1,000,000,000+' 
                                       or apps[5]== '500,000,000+'
                                       or apps[5]== '100,000,000+'):
        print(apps[0], ':',apps[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

Removing apps with installs above 100 M + will bring down the category avg install number down drastically. 

In [27]:
under_100m = []

for app in google_free:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    n_installs = float(n_installs)
    if app[1] == 'COMMUNICATION' and n_installs<100000000:
        under_100m.append(n_installs)
        
sum(under_100m)/len(under_100m)

3603485.3884615386

**Exploring another Category**

In [28]:
for app in google_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+



However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [29]:
for app in google_free:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H


This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

# Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.


Similar process can be taken to analyze and slice through IOS based apps

In [30]:
#example to revise nested loop confusion

some_string = ['First','Second']
some_int = [1,2,3,4]

for e in some_string:
    print(e)
    for z in some_int:
        print(z)
    x = sum(some_int)
    print(x)

First
1
2
3
4
10
Second
1
2
3
4
10
